src/inscription.rs

```diff
@@ -0,0 +1,377 @@
+    let mut witness = witness.to_vec();
+
+    if witness.is_empty() {
+      return Err(Error::EmptyWitness);
+    }
+
+    if witness.len() > 1
+      && witness
+        .last()
+        .and_then(|element| element.first().map(|byte| *byte == TAPROOT_ANNEX_PREFIX))
+        .unwrap_or(false)
+    {
+      witness.pop();
+    }
+
+    if witness.len() == 1 {
+      return Err(Error::KeyPathSpend);
+    }
+
+    // remove control block
+    witness.pop().unwrap();
+
+    // extract script
+    let script = Script::from(witness.pop().unwrap());
+
+    let instructions = script
+      .instructions()
+      .collect::<Result<Vec<Instruction>, script::Error>>()
+      .map_err(Error::Script)?;
+
+    InscriptionParser {
+      next: 0,
+      instructions,
+    }
+    .parse_script()
+  }
+
+  fn parse_script(mut self) -> Result<Inscription> {
+    loop {
+      let next = self.advance()?;
+
+      if next == Instruction::PushBytes(&[]) {
+        if let Some(inscription) = self.parse_inscription()? {
+          return Ok(inscription);
+        }
+      }
+    }
+  }
+
+  fn advance(&mut self) -> Result<Instruction<'a>> {
+    let next = self
+      .instructions
+      .get(self.next)
+      .ok_or(Error::NoInscription)?;
+    self.next += 1;
+    Ok(next.clone())
+  }
+
+  fn parse_inscription(&mut self) -> Result<Option<Inscription>> {
+    if self.advance()? == Instruction::Op(opcodes::all::OP_IF) {
+      let content = self.advance()?;
+
+      let content = if let Instruction::PushBytes(bytes) = content {
+        str::from_utf8(bytes).map_err(Error::Utf8Decode)?
+      } else {
+        return Err(Error::InvalidInscription);
+      };
+
+      if self.advance()? != Instruction::Op(opcodes::all::OP_ENDIF) {
+        return Err(Error::InvalidInscription);
+      }
+
+      return Ok(Some(Inscription(content.to_string())));
+    }
+
+    Ok(None)
+  }
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::{Ordinal, OutPoint, Sequence, Transaction, TxIn, VecDeque};
+
+  #[test]
+  fn empty() {
+    assert_eq!(
+      InscriptionParser::parse(&Witness::new()),
+      Err(Error::EmptyWitness)
+    );
+  }
+
+  #[test]
+  fn ignore_key_path_spends() {
+    assert_eq!(
+      InscriptionParser::parse(&Witness::from_vec(vec![vec![]])),
+      Err(Error::KeyPathSpend),
+    );
+  }
+
+  #[test]
+  fn ignore_key_path_spends_with_annex() {
+    assert_eq!(
+      InscriptionParser::parse(&Witness::from_vec(vec![vec![], vec![0x50]])),
+      Err(Error::KeyPathSpend),
+    );
+  }
+
+  #[test]
+  fn ignore_unparsable_scripts() {
+    assert_eq!(
+      InscriptionParser::parse(&Witness::from_vec(vec![vec![0x01], vec![]])),
+      Err(Error::Script(script::Error::EarlyEndOfScript)),
+    );
+  }
+
+  #[test]
+  fn no_inscription() {
+    assert_eq!(
+      InscriptionParser::parse(&Witness::from_vec(vec![Script::new().into_bytes(), vec![]])),
+      Err(Error::NoInscription),
+    );
+  }
+
+  #[test]
+  fn valid() {
```

Comment:
Add a test that when there are two valid inscriptions in the same witness, the first inscription is used.